runtime4/misc.c

```diff
@@ -252,11 +252,20 @@ CAMLprim value caml_atomic_load(value ref)
   return Field(ref, 0);
 }
 
-CAMLprim value caml_atomic_cas(value ref, value oldv, value newv)
+CAMLprim value caml_atomic_compare_exchange(value ref, value oldv, value newv)
 {
   value* p = Op_val(ref);
   if (*p == oldv) {
     caml_modify(p, newv);
+    return oldv;
+  } else {
+    return *p;
+  }
+}
+
+CAMLprim value caml_atomic_cas(value ref, value oldv, value newv)
+{
+  if (caml_atomic_compare_exchange(ref, oldv, newv) == oldv) {
     return Val_int(1);
```

Comment:
not changed here, but... should this have been using `Val_bool` instead?

Reply:
sure